src/diagnostics.rs

```diff
@@ -157,6 +157,35 @@ pub fn report_error<'tcx, 'mir>(
         }
     };
 
+    let mut stacktrace = ecx.generate_stacktrace();
+    match ecx.machine.backtrace_style {
+        BacktraceStyle::Off => {
+            // Retain one frame so that we can print a span for the error itself
+            stacktrace.truncate(1);
+        }
+        BacktraceStyle::Short => {
+            // This is part of the logic that `std` uses to select the relevant part of a
+            // backtrace. But here, we only look for __rust_begin_short_backtrace, not
+            // __rust_end_short_backtrace because the end symbol comes from a call to the default
+            // panic handler.
+            stacktrace = stacktrace
+                .into_iter()
+                .take_while(|e| {
+                    let def_id = e.instance.def_id();
+                    let path = ecx.tcx.tcx.def_path(def_id).to_string_no_crate_verbose();
+                    !path.contains("__rust_begin_short_backtrace")
+                })
+                .collect::<Vec<_>>();
+            // After we prune frames from the bottom, there are a few left that are part of the
+            // Rust runtime. So we remove frames until we get to a local symbol, which should be
+            // main or a test.
+            while stacktrace.last().map_or(false, |e| !e.instance.def_id().is_local()) {
```

Comment:
One potential concern here is what happens when the error actually points to inside the Rust runtime. In general it's probably okay to require the flag in that case, but can you at least make sure that this will never leave the stacktrace empty? (Or maybe you have a better idea for how to cover that case?)

Reply:
I'm having trouble thinking of how that would happen in normal execution. Are you imagining running Miri and finding a bug in the `cargo test` test harness? I'm pretty sure in that scenario, `RUST_BACKTRACE=1` breaks completely too.

I suspect the note about full backtraces exists to cover this case.

Reply:
Yeah, Miri has found UB in the standard library in the past, so that might happen again.

Maybe a suitable heuristic would be: If there is *any* local frame, then prune the non-local ones from the bottom? That will ensure we do something reasonable for "entirely non-local" backtraces.

Reply:
I think if we end up in this situation, the best thing to do is not prune at all. The only thing that all this pruning does is hide the runtime, and if that's your bug territory, bad.

We always need the first stage of pruning (if we're pruning at all), because the `cargo miri test` backtraces pass through a local frame on their way to `__rust_begin_short_backtrace`.

Reply:
> We always need the first stage of pruning (if we're pruning at all), because the cargo miri test backtraces pass through a local frame on their way to __rust_begin_short_backtrace.

That's okay; in the cases I worry about I think there will be no __rust_begin_short_backtrace in the backtrace at all.